### PR TITLE
LifeCycleAssesment_oM: Add new LCA result classes

### LINKED ISSUE #1448: LifeCycleAssesment_oM: Breakdown results per EPD so we can know totals per materials
#### Description:
Currently total Metrics are given per element, but not per material. For a lot of LCA analysis we need to know the quantities (such as GWP) per material, and an EvaluationMethod that provides a result per epd is essential to these kinds of analysis.

For example, a Wall with a Construction with 2 layers made up of 2 different materials is evaluated. Currently, the Evalutaion returns a single value. This new method would return 2 Results, one for each material with their respective quantities. THe current structure of the evaluation methods means this could be easily added by calling the helper evaluation methods per EPD, and collecting these as separate results rather than combining them into a single result.

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #1448

Adding new LCA result classes:
- ElementResult - can replace the current result class if this PR is accepted. Stores information about how to identify the result as well as total quantity and a breakdown of quantity evaluated per material.
- MaterialResult - New result class, storing information about quantity per material. Can be per element as part of ElementResult or total for a particular material.

Adding as new classes as a first step. If agreed we could replace the current ResultClasses in the oM.

### CODE REVIEW

File: LifeCycleAssessment_oM/Results/MaterialResult.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+namespace BH.oM.LifeCycleAssessment.Results
+{
+    [Description("Result for a material/epd evaluated..")]
+    public class MaterialResult : IResult, IImmutable
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("Name fo the physical material evaluated.")]
```
Comment (kayleighhoude) [inline LifeCycleAssessment_oM/Results/MaterialResult.cs:17]: of*

File: LifeCycleAssessment_oM/Results/MaterialResult.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        public virtual string MaterialName { get; }
+
+        [Description("Name fo the EPD material evaluated.")]
```
Comment (kayleighhoude) [inline LifeCycleAssessment_oM/Results/MaterialResult.cs:20]: of*

File: LifeCycleAssessment_oM/Results/MaterialResult.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        public virtual string EPDName { get; }
+
+        [Description("The total quantity of the specified Field within the object.")]
+        public virtual double Quantity { get; }
+
+        [Description("Phase of life abbreviation for the scope of the EPD. A single EnvironmentalMetric can contain either a single Phase or a list of Phases i.e. A1, A2, A3.")]
+        public virtual IReadOnlyList<LifeCycleAssessmentPhases> Phases { get; }
+
+        [Description("The EPD Field selected for evaluation.")]
```
Comment (kayleighhoude) [inline LifeCycleAssessment_oM/Results/MaterialResult.cs:29]: more verbose: Environmental Product Declaration

File: LifeCycleAssessment_oM/Results/ElementResult.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        [Description("Category of the object this result was generated from, e.g. Beam or Wall")]
+        public virtual ObjectCategory Category { get; } = ObjectCategory.Undefined;
+
+        [Description("Phase of life abbreviation for the scope of the EPD. A single EnvironmentalMetric can contain either a single Phase or a list of Phases i.e. A1, A2, A3.")]
+        public virtual IReadOnlyList<LifeCycleAssessmentPhases> Phases { get; } = new List<LifeCycleAssessmentPhases>();
+
+        [Description("The total quantity of the specified Field within the object.")]
+        public virtual double Quantity { get; set; }
+
+        [Description("The EPD Field selected for evaluation.")]
```
Comment (kayleighhoude) [inline LifeCycleAssessment_oM/Results/ElementResult.cs:32]: More verbose: Environmental Product Declaration

**Review by kayleighhoude — CHANGES_REQUESTED** [review]
Finicky comments on spelling, otherwise happy to give an approving review

**Comment by IsakNaslundBh** [comment]
THanks @kayleighhoude . Have updated the descriptions mentioned (and a few more). Glad for a second review! :)

File: LifeCycleAssessment_oM/Results/ElementResult.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
+﻿using BH.oM.Analytical.Results;
```
Comment (FraserGreenroyd) [inline LifeCycleAssessment_oM/Results/ElementResult.cs:1]: Copyright header is required

File: LifeCycleAssessment_oM/Results/MaterialResult.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
+﻿using BH.oM.Analytical.Results;
```
Comment (FraserGreenroyd) [inline LifeCycleAssessment_oM/Results/MaterialResult.cs:1]: Copyright header needed

File: LifeCycleAssessment_oM/Results/MaterialResult.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("Name of the physical material evaluated.")]
+        public virtual string MaterialName { get; }
+
+        [Description("Name of the Environmental Product Declaration evaluated.")]
+        public virtual string EPDName { get; }
```
Comment (FraserGreenroyd) [inline LifeCycleAssessment_oM/Results/MaterialResult.cs:21]: Avoiding acronyms and matching `EnvironmentalProductDeclercationField`, this should be:
```suggestion
        public virtual string EnvironmentalProductDeclarationName { get; }
```

File: LifeCycleAssessment_oM/Results/ElementResult.cs (hunk 4 above)
Comment (FraserGreenroyd) [reply LifeCycleAssessment_oM/Results/ElementResult.cs:32]:
```suggestion
        [Description("The Environmental Product Declaration Field selected for evaluation.")]
```

Conversation was marked as resolved @IsakNaslundBh so not sure if it's just been missed from a push perhaps?

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Minor comments 😄

File: LifeCycleAssessment_oM/Results/ElementResult.cs (hunk 4 above)
Comment (IsakNaslundBh) [reply LifeCycleAssessment_oM/Results/ElementResult.cs:32]: You are right. Updated it in one place but not the other

**Review by kayleighhoude — DISMISSED** [review]
Approved per recent changes

**Comment by michaelhoehn** [comment]
@BHoMBot check required

**Comment by michaelhoehn** [comment]
@BHoMBot check copyright-compliance

**Review by FraserGreenroyd — APPROVED** [review]
This body of work has been reviewed at length by @michaelhoehn who has talked me through it today. Owing to the upcoming feature freeze and desire to have this in for the 6.0 beta, and with @IsakNaslundBh currently unavailable, I have agreed with @michaelhoehn to merge this PR and the supporting body of work to put the feature in, with the understanding that he and @IsakNaslundBh will rectify any bugs in the next sprint, and provide unit tests for the engine methods and LifeCycleAssessment_Toolkit as appropriate.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 9834606863

**Pull request merged**